### DISCUSSION TITLE
Can we use the github app or oauth app instead of github PAT to generate the RUNNER_TOKEN?

### DISCUSSION THREAD

**Discussion by michaelsync** [community/community#24743] [Apps, API and Webhooks]
<p>Can we use the github app or oauth app instead of github PAT to generate the RUNNER_TOKEN? The problem with PAT is that none of our Github organization admins want to give the PAT.</p>
<p>I am trying to register the self-hosted runner (Docker container) at the organization level using <a href="https://docs.github.com/en/free-pro-team@latest/rest/reference/actions#create-a-registration-token-for-an-organization" rel="noopener nofollow ugc">this API</a>. (I am following <a href="https://sanderknape.com/2020/03/self-hosted-github-actions-runner-kubernetes/" rel="noopener nofollow ugc">this article</a> except for the k8s part.). If I use the personal access token of the Github organization admin, it works. But we don’t want to use it from the docker runner since it’s the “personal” access token.</p>
<p>I have looked at Oauth App device flow but it needs some manual things like entering the code and etc.</p>

**Comment by ghost** [comment, marked as answer]
<p>Here is an overview of the entire process for creating a self-hosted registration token using an App:</p>
<ul>
<li>
<p>Create a GitHub App</p>
</li>
<li>
<p>Create a PEM for that app: <a href="https://docs.github.com/en/free-pro-team@latest/developers/apps/authenticating-with-github-apps#generating-a-private-key">https://docs.github.com/en/free-pro-team@latest/developers/apps/authenticating-with-github-apps#generating-a-private-key</a></p>
</li>
<li>
<p>Install that app onto your account</p>
</li>
<li>
<p>Based on <a href="https://docs.github.com/en/free-pro-team@latest/developers/apps/authenticating-with-github-apps#authenticating-as-a-github-app">Auth as a GitHub App</a>, copy the ruby script and generate a JWT token</p>
</li>
<li>
<p>Retrieve the installation ID using the JWT with auth as Bearer<br>
<a href="https://docs.github.com/en/free-pro-team@latest/rest/reference/apps#list-installations-for-the-authenticated-app">https://docs.github.com/en/free-pro-team@latest/rest/reference/apps#list-installations-for-the-authenticated-app</a></p>
</li>
<li>
<p>Create an access token for the installation<br>
<a href="https://docs.github.com/en/free-pro-team@latest/rest/reference/apps#create-an-installation-access-token-for-an-app">https://docs.github.com/en/free-pro-team@latest/rest/reference/apps#create-an-installation-access-token-for-an-app</a></p>
</li>
<li>
<p>Using that access token, create a runner registration token<br>
<a href="https://docs.github.com/en/free-pro-team@latest/rest/reference/actions#create-a-registration-token-for-a-repository">https://docs.github.com/en/free-pro-team@latest/rest/reference/actions#create-a-registration-token-for-a-repository</a></p>
</li>
</ul>

**Comment by michaelsync** [comment]
<p>Thanks, <a class="mention" href="/github-support">@github-support</a>.<br>
I created the Github app in my organization. So I can get the app id.</p>
<p>Do you have a bash script instead of Ruby to genereate the JWT?</p>
<p>I tried to use this one <a href="https://gist.github.com/michaelsync/4ad968e96631dbb9638e57de29b6b937" rel="noopener nofollow ugc">https://gist.github.com/michaelsync/4ad968e96631dbb9638e57de29b6b937</a> but didn’t work. I am getting this error. My bash works fine with PAT (both repo level or org level).</p>
<pre><code class="lang-auto">{"message":"Not Found","documentation_url":"https://docs.github.com/rest"}
Response status code does not indicate success: 404 (Not Found).</code></pre>

**Comment by michaelsync** [comment]
<p>No worries. I found a way to do it. Thanks.</p>

**Comment by apostoins** [comment]
<p>Have you found the way to do it? Could you share?</p>

**Comment by michaelsync** [comment]
<p><a class="mention" href="/apostoins">@apostoins</a> It has been a while. I think I have used this <a href="https://github.com/michaelsync/github-actions-self-hosted-runner-container/blob/ae6c023f0985d1b5d35731f592d2abe577a86b98/entrypoint.sh" class="inline-onebox" rel="noopener nofollow ugc">github-actions-self-hosted-runner-container/entrypoint.sh at ae6c023f0985d1b5d35731f592d2abe577a86b98 · michaelsync/github-actions-self-hosted-runner-container · GitHub</a> or one of those files from that repo.</p>
<p>I was trying to register the runner at the organization level or repo. I managed to get it working but I need to clean up the files and arrange it properly if I get some times.</p>
